rare/lgndr/manager.py

```diff
@@ -0,0 +1,175 @@
+
+            perc = (processed_chunks / num_chunk_tasks) * 100
+            runtime = time.time() - s_time
+            total_avail = len(self.sms)
+            total_used = (num_shared_memory_segments - total_avail) * (self.analysis.biggest_chunk / 1024 / 1024)
+
+            try:
+                average_speed = processed_chunks / runtime
+                estimate = (num_chunk_tasks - processed_chunks) / average_speed
+            except ZeroDivisionError:
+                average_speed = estimate = 0
+
+            # send status update to back to instantiator (if queue exists)
+            if self.status_queue:
+                try:
+                    self.status_queue.put(UIUpdate(
+                        progress=perc,
+                        runtime=round(runtime),
+                        estimated_time_left=round(estimate),
+                        processed_chunks=processed_chunks,
+                        chunk_tasks=num_chunk_tasks,
+                        total_downloaded=total_dl,
+                        total_written=total_write,
+                        cache_usage=total_used,
+                        active_tasks=self.active_tasks,
+                        download_speed=dl_speed,
+                        download_decompressed_speed=dl_unc_speed,
+                        write_speed=w_speed,
+                        memory_usage=total_used * 1024 * 1024,
+                        read_speed=r_speed,
+                    ), timeout=1.0)
+                except Exception as e:
+                    self.log.warning(f'Failed to send status update to queue: {e!r}')
+
+            time.sleep(self.update_interval)
+
+        for i in range(self.max_workers):
+            self.dl_worker_queue.put_nowait(TerminateWorkerTask())
+
+        self.log.info('Waiting for installation to finish...')
+        self.writer_queue.put_nowait(TerminateWorkerTask())
+
+        writer_p.join(timeout=10.0)
+        if writer_p.exitcode is None:
+            self.log.warning(f'Terminating writer process, no exit code!')
+            writer_p.terminate()
+
+        # forcibly kill DL workers that are not actually dead yet
+        for child in self.children:
+            if child.exitcode is None:
+                child.terminate()
+
+        # make sure all the threads are dead.
+        for t in self.threads:
+            t.join(timeout=5.0)
+            if t.is_alive():
+                self.log.warning(f'Thread did not terminate! {repr(t)}')
+
+        # clean up resume file
+        if self.resume_file:
+            try:
+                os.remove(self.resume_file)
+            except OSError as e:
+                self.log.warning(f'Failed to remove resume file: {e!r}')
+
+        # close up shared memory
+        self.shared_memory.close()
+        self.shared_memory.unlink()
+        self.shared_memory = None
+
+        self.log.info('All done! Download manager quitting...')
+        # finally, exit the process.
+        exit(0)
```

Comment:
Why the explicit exit, actually?

Reply:
I don't know why that was there in the first place. Might have to do with `multiprocessing`, have to investigate.